llvm/include/llvm/BinaryFormat/DXContainer.h

```diff
@@ -546,13 +560,75 @@ struct ProgramSignatureElement {
 static_assert(sizeof(ProgramSignatureElement) == 32,
               "ProgramSignatureElement is misaligned");
 
+struct RootConstants {
+  uint32_t ShaderRegister = 0;
+  uint32_t RegisterSpace = 0;
+  uint32_t Num32BitValues = 0;
+
+  RootConstants() = default;
+
+  void swapBytes() {
+    sys::swapByteOrder(ShaderRegister);
+    sys::swapByteOrder(RegisterSpace);
+    sys::swapByteOrder(Num32BitValues);
+  }
+};
+
+struct RootParameter {
+  dxbc::RootParameterType ParameterType;
+  union {
+    dxbc::RootConstants Constants;
+  };
+  dxbc::ShaderVisibilityFlag ShaderVisibility;
+
+  RootParameter() {
+    Constants = RootConstants();
+    ParameterType = dxbc::RootParameterType::Empty;
+    ShaderVisibility = dxbc::ShaderVisibilityFlag::Empty;
+  }
+
+  void swapBytes() {
+
+    sys::swapByteOrder(ParameterType);
+    sys::swapByteOrder(ShaderVisibility);
+    switch (ParameterType) {
+    case RootParameterType::Constants32Bit:
+      Constants.swapBytes();
+      break;
+    case RootParameterType::Empty:
+      llvm_unreachable("invalid value for ParameterType");
+      break;
+    }
+  }
+};
+
+struct RootSignatureHeader {
+  uint32_t Version = 2;
+  uint32_t Flags = 0;
```

Comment:
This should not be default initialized. The BinaryFormat library's purpose is to define the structures used for both writing _and_ reading. When reading you definitely don't want to default initialize data members.